packages/massa-proto-cli/src/ts-gen.ts

```diff
@@ -214,24 +242,109 @@ let callSC: (address: string, funcName: string, binArguments: Uint8Array, maxCoi
  * 
  ${documentationArgs.slice(1)}
  *
- * @returns {IEvent[]} A promise that resolves to an array of events generated during the execution of ${protoFile.funcName}.
+ * @returns {Promise<OperationOutputs>} A promise that resolves to an object which contains the outputs and events from the call to ${protoFile.funcName}.
  */
- export async function ${protoFile.funcName}(${args}): Promise<IEvent[]> {
+ export async function ${protoFile.funcName}(${args}): Promise<OperationOutputs> {
   ${checkUnsignedArgs}
 
   ${argsSerialization}
 
   // Send the operation to the blockchain and retrieve its outputs
   return (
-    await getEvents(
+    await extractOutputsAndEvents(
+      '${contractAddress}',
+      '${protoFile.funcName}',
+      serializedArgs,
+      coins,
+      '${returnType[protoFile.resType]}',
+    )
+  );
+}
+
+async function extractOutputsAndEvents(
+  contractAddress: string, 
+  functionName: string, 
+  args: Uint8Array, 
+  coins: bigint, 
+  returnType: string
+  ): Promise<OperationOutputs> {
+
+  let events: IEvent[] = [];
+
+  // try to call the Smart Contract
+  try{
+    events = await getEvents(
       await callSC(
-        '${contractAddress}',
-        '${protoFile.funcName}',
-        serializedArgs,
+        contractAddress,
+        functionName,
+        args,
         coins,
       ) as TransactionDetails
     )
-  );
+  }
+  catch (err) { // if the call fails, return the error
+    return {
+      events: events,
+      isError: true,
+      error: err,
+    } as OperationOutputs;
+  }
+
+  // if the call is successful, retrieve the outputs from the events
+  let rawOutput: string | null = null;
+  for (let event of events) {
+    if (event.data.slice(0, OUTPUTS_PREFIX.length) == OUTPUTS_PREFIX) {
+      rawOutput = event.data.slice(OUTPUTS_PREFIX.length);
+      // remove the event from the list
+      events.splice(events.indexOf(event), 1);
+      break;
+    }
+  }
+
+  // check the output and return the result
+  if (rawOutput === null && returnType !== 'void') {
+    return {
+      events: events,
+      isError: true,
+      error: 'No outputs found. Expected type: ' + returnType,
+    } as OperationOutputs;
+  }
+  else if(rawOutput === null && returnType === 'void') {
```

Comment:
else is useless as you have a return in the if branch

Reply:
sure